pkg/services/ngalert/state/historian/annotation.go

```diff
@@ -56,19 +55,19 @@ func (h *AnnotationStateHistorian) RecordState(ctx context.Context, rule *ngmode
 			return
 		}
 
-		query := &models.GetDashboardQuery{
-			Uid:   dashUid,
-			OrgId: rule.OrgID,
-		}
-
-		err = h.dashboards.GetDashboard(ctx, query)
+		dashID, err := h.dashboards.getId(ctx, rule.OrgID, dashUid)
 		if err != nil {
 			h.log.Error("error getting dashboard for alert annotation", "dashboardUID", dashUid, "alertRuleUID", rule.UID, "err", err.Error())
 			return
 		}
 
+		if dashID == dashboardNotFound {
```

Comment:
How come we check a magic value instead of `dashboards.ErrDashboardNotFound`?

Reply:
Firstly we want to be sure and cache failed lookups. That way, rules with a badly configured dashboard ID don't circumvent the caching.

I felt it was cleaner logic-wise to have the signifier for "this query failed" be the same underlying type as the rest of the data being cached. If we end up storing mixed types in the cache, we have to use reflection each time we read in order to figure out what's actually in the cache value.

I agree it does pollute the caller experience a bit though. I'm not opposed to switching to reflection if we think it's a big enough improvement, WDYT?

Reply:
After moving to singleflight, this was less of a problem than it was prior!